src/VaryingAttributes.cpp

```diff
@@ -473,16 +473,54 @@ void prune_varying_attributes(Stmt loop_stmt, std::map<std::string, Expr>& varyi
     }
 }
 
-// This visitor changes the type of variables tagged with .varying to float, and
-// then propagates the float expression up through the expression, casting where
-// necessary.
-class CastVaryingVariablesToFloat : public IRMutator {
+// This visitor changes the type of variables tagged with .varying to float,
+// since GLSL will only interpolate floats. In the case that the type of the
+// varying attribute was integer, the interpolated float value is snapped to the
+// integer grid and cast to the integer type. This case occurs with coordinate
+// expressions where the integer loop variables are manipulated without being
+// converted to floating point. In other cases, like an affine transformation of
+// image coordinates, the loop variables are cast to floating point within the
+// interpolated expression.
+class CastVaryingVariables : public IRMutator {
 protected:
     using IRMutator::visit;
 
     virtual void visit(const Variable *op) {
         if ((ends_with(op->name, ".varying")) && (op->type != Float(32))) {
-            expr = Variable::make(Float(32), op->name);
+            // The incoming variable will be float type because GLSL only
+            // interpolates floats
+            Expr v = Variable::make(Float(32), op->name);
+
+            // If the varying attribute expression that this variable replaced
+            // was integer type, snap the interpolated floating point variable
+            // back to the integer grid.
+            expr = Cast::make(op->type, floor(v + 0.5f));
+        } else {
+            // Otherwise, the variable keeps its float type.
+            expr = op;
+        }
+    }
+};
+
+// This visitor casts the named variables to float, and then propagates the
+// float type through the expression. The variable is offset by 0.5f
+class CastVariablesToFloatAndOffset : public IRMutator {
+protected:
+    using IRMutator::visit;
+
+    virtual void visit(const Variable *op) {
+
+        // Check to see if the variable matches a loop variable name
+        if (std::find(names.begin(), names.end(), op->name) != names.end()) {
+            // This case is used by integer type loop variables. They are cast
+            // to float and offset.
+            expr = Sub::make(Cast::make(float_type(op), op), 0.5f);
```

Comment:
Yes, that would be fine. In the past I had more frequently used the make methods to construct IR during lowering.

This section of code was cut and paste unchanged into CastVariablesToFloatAndOffset since the parent class it previously belonged to no longer exists.